plugins/outputs/sql/sql.go

```diff
@@ -201,28 +211,95 @@ func (p *SQL) generateInsert(tablename string, columns []string) string {
 		strings.Join(placeholders, ","))
 }
 
+func (p *SQL) createTable(metric telegraf.Metric) error {
+	tablename := metric.Name()
+	if _, found := p.tables[tablename]; !found && !p.tableExists(tablename) {
+		stmt := p.generateCreateTable(metric)
+		if _, err := p.db.Exec(stmt); err != nil {
+			return fmt.Errorf("creating table failed: %w", err)
+		}
+		p.tables[tablename] = make(map[string]bool)
+		if err := p.updateTableCache(tablename); err != nil {
+			return fmt.Errorf("updating table cache failed: %w", err)
+		}
+	}
+	return nil
+}
+
+func (p *SQL) createColumn(tablename, column, columnType string) error {
+	// modifying the table schema is opt-in
+	if p.TableUpdateTemplate == "" {
+		return nil
+	}
+
+	// update table cache with the latest table schema
+	// some servers may lack support for "IF NOT EXIST"
+	// clause in which case, we don't want to burden users
+	// with writing complex "IF NOT EXIST" workarounds
+	if !p.tables[tablename][column] {
+		if err := p.updateTableCache(tablename); err != nil {
+			return fmt.Errorf("updating table cache failed: %w", err)
+		}
+	}
+
+	if !p.tables[tablename][column] {
+		createColumn := p.generateAddColumn(tablename, column, columnType)
+		if _, err := p.db.Exec(createColumn); err != nil {
+			return fmt.Errorf("creating column failed: %w", err)
+		}
+		if err := p.updateTableCache(tablename); err != nil {
+			return fmt.Errorf("updating table cache failed: %w", err)
+		}
+	}
+
+	return nil
+}
+
 func (p *SQL) tableExists(tableName string) bool {
 	stmt := strings.ReplaceAll(p.TableExistsTemplate, "{TABLE}", quoteIdent(tableName))
 
 	_, err := p.db.Exec(stmt)
 	return err == nil
 }
 
+func (p *SQL) updateTableCache(tablename string) error {
+	// modifying the table schema is opt-in
+	if p.TableUpdateTemplate == "" {
+		return nil
+	}
```

Comment:
Can we please move the condition to the caller place to make clear that we only update if opted-in!? It's more a code reading thing than a functional change...